### PR TITLE
chore(request-response): remove deprecated items

### PR DESCRIPTION
**Pull request by umgefahren**
## Description

Related: https://github.com/libp2p/rust-libp2p/issues/3647.

## Notes & open questions

Should be merged with next breaking change to RequestResponse

## Change checklist

- [x] I have performed a self-review of my own code
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
We've now opened the merge window for breaking changes so this is good to go in once the merge conflicts are resolved.

**Comment by umgefahren** [comment]
I have resolved the merge conflicts.

File: protocols/request-response/src/lib.rs
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs]:
```suggestion
mod codec;
mod handler;
```

We want these to be private!

**Review by thomaseizinger — CHANGES_REQUESTED** [review]

**Review by thomaseizinger — APPROVED** [review]

**Comment by aakoshh** [comment]
I have code that relies on pattern matching on the `request_response::handler::Event` enum to look at the request (for [rate limiting purposes](https://github.com/consensus-shipyard/ipc/pull/90)). How can I update that code now that the `handler` module is private?

**Comment by thomaseizinger** [comment]
> I have code that relies on pattern matching on the `request_response::handler::Event` enum to look at the request (for [rate limiting purposes](https://github.com/consensus-shipyard/ipc/pull/90)). How can I update that code now that the `handler` module is private?

`Event` contains a `Message` and `Message` is public. Where are you getting the `handler::Event` from?

**Comment by aakoshh** [comment]
> `Event` contains a `Message` and `Message` is public. 

True, but that that's `request_response::Event`, and here I am handling `request_response::handler::Event` which is different. `Message::Request` has different fields.

My use case was that I wrapped the `libp2p-bitswap` into a behaviour of my own (packaged it up with others to do peer discovery and gossiping as well), and I wanted to have some rate limiting per peer ID to limit the amount of data that can be served over Bitswap to reasonable amounts. I didn't want to modify the library itself, and was happy to find a (hacky) way to do rate limiting in my wrappers.

The structure looked like this (sorry for the wall of code):

```rust
pub struct Behaviour<P: StoreParams> {
    inner: Bitswap<P>,
    /// Limit the amount of data served by remote address.
    rate_limiter: RateLimiter<Multiaddr>,
    rate_limit: RateLimit,
    outbox: VecDeque<Event>,
}

impl<P: StoreParams> NetworkBehaviour for Behaviour<P> {
    type ConnectionHandler = <Bitswap<P> as NetworkBehaviour>::ConnectionHandler;
    type OutEvent = Event;

    fn on_connection_handler_event(
        &mut self,
        peer_id: PeerId,
        connection_id: ConnectionId,
        event: <<Self::ConnectionHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
    ) {
        match event {
            RequestResponseHandlerEvent::Request {
                request_id,
                request,
                sender,
            } if self.has_rate_limits() => {
                if !self.check_rate_limit(&peer_id, &request.cid) {
                    warn!("rate limiting {peer_id}");
                    stats::CONTENT_RATE_LIMITED.inc();
                    return;
                }
                // We need to hijack the response channel to record the size, otherwise it goes straight to the handler.
                let (tx, rx) = libp2p::futures::channel::oneshot::channel();
                let event = RequestResponseHandlerEvent::Request {
                    request_id,
                    request,
                    sender: tx,
                };

                self.inner
                    .on_connection_handler_event(peer_id, connection_id, event);

                let forward = Event::BitswapForward {
                    peer_id,
                    response_rx: rx,
                    response_tx: sender,
                };
                self.outbox.push_back(forward);
            }
            _ => self
                .inner
                .on_connection_handler_event(peer_id, connection_id, event),
        }
    }
```

What's happening is that before I forwarded the request to the `inner` Bitswap behaviour, I inspected whether the event going from the connection handler to the `inner` contained a `BitswapRequest` (not a response, which I didn't want to rate limit). If so, I had to do this elaborate dance of setting up my own channels that allowed me to inspect the response I sent back to the peer and count the number of bytes they got served. 

I believe the `Message` you mention is a `GeneratedEvent` raised by the `request_response::Behaviour` and it's handled by the `Bitswap` behaviour itself, so I never see it, and couldn't inject my bookkeeping. 

So, my use case partially boils down to: how can I selectively decide not to forward a connection handler event to the inner behaviour if I have no way to inspect it at this level?

**Comment by thomaseizinger** [comment]
Oh I see what you are doing. We don't really support that as a usecase. Everything but the `NetworkBehaviour` and its public functions is not considered public API. We do that to make sure we can make changes without having to publish them as minor versions.

I don't know the structure of your bitswap implementation but the way I would approach it is to expose events that you can hook into and decide to drop a request if the limit is exceeded, similar to how our request-response implementation exposes an event for an incoming request.

Or you modify the bitswap code to have first-class support for rate limits (which also seems reasonable).

Our request-response module is designed for simple usecases and purposely contrained in its API. If it doesn't work for you, I'd recommend dropping it and writing your own `NetworkBehaviour` and `ConnectionHandler`. This applies mostly to `libp2p-bitswap` in this case.

**Comment by aakoshh** [comment]
Okay, thanks @thomaseizinger, I'll add support for some kind of rate limiting to our fork of `libp2p-bitswap`. 

I understand how you wanted to limit what `request_response` exposes. I agree that what I did relied on a particular implementation of the event, and while I was using only public fields, and nobody got hurt, it probably wasn't what you had in mind originally. 

It's a little bit strange now that all I can really do (and _have_ to do) is forward the `on_connection_handler_event`, so while I can do shortcut logic, AFAIK I can't tell even if this is a request or a response, so my choices are rather limited, and this seems to be a choice made by the `request_response` protocol, not a limitation of the `NetworkBehaviour` itself. Oh well.

**Comment by thomaseizinger** [comment]
> Okay, thanks @thomaseizinger, I'll add support for some kind of rate limiting to our fork of `libp2p-bitswap`.
> 
> I understand how you wanted to limit what `request_response` exposes. I agree that what I did relied on a particular implementation of the event, and while I was using only public fields, and nobody got hurt, it probably wasn't what you had in mind originally.
> 
> It's a little bit strange now that all I can really do (and _have_ to do) is forward the `on_connection_handler_event`, so while I can do shortcut logic, AFAIK I can't tell even if this is a request or a response, so my choices are rather limited, and this seems to be a choice made by the `request_response` protocol, not a limitation of the `NetworkBehaviour` itself. Oh well.

To be precise, the problem you are experiencing is a combination of manually composing behaviours (which is fine in itself) and the fact that you want to intercept the _handler_ event of a "4th" party behaviour (you don't control `libp2p-bitswap` and are dealing with something that it depends on).

What we do want to support is composing _behaviours_ manually to add extra logic. We do that within `libp2p-rendezvous` for example which uses `libp2p-request-response`. What we don't really support is intercepting handler events as those are considered internal to a particular module.

Essentially, what you are trying to do is make a modified version of bitswap that re-composes its parts for new functionality. If it is a design-goal of `libp2p-bitswap` to allow that, then the proper way to handle this would be to encapsulate `libp2p-request-response` entirely and not use it in its public API. Hence, I'd argue it is `libp2p-bitswap` that needs to change, not `libp2p-request-response`! :)

**Comment by aakoshh** [comment]
Thank you for the pointers and explaining the rationale. It's not completely clear to me why a behaviour can re-map output events of an inner behaviour but not modify the inputs of its handlers, but it good to know this is the intention. For example I don't know how `libp2p-bitswap` could completely encapsulate `libp2p-request-response` if you can't feed it the handler events. I'll check out `libp2p-rendezvous` :+1: 

To be fair `libp2p-bitswap` also didn't have this available in its public API, I had to [ask them](https://github.com/ipfs-rust/libp2p-bitswap/pull/44) to make the request and response types public. It seemed there was no harm in it.

**Comment by thomaseizinger** [comment]
> For example I don't know how `libp2p-bitswap` could completely encapsulate `libp2p-request-response` if you can't feed it the handler events. I'll check out `libp2p-rendezvous` 👍

If `libp2p-bitswap` only composes `libp2p-request-response` and forwards all events, then it doesn't need to name the handler events. Essentially, when you compose another behaviour, the only things you should customize is:
- what you do with the event type when you poll the inner behaviour
- which public APIs you call on it

This is how rendezvous is implemented for example: https://github.com/libp2p/rust-libp2p/blob/master/protocols/rendezvous/src/client.rs.

**Pull request merged**
